ArduPlane/mode_autoland.cpp

```diff
@@ -129,30 +139,97 @@ void ModeAutoLand::update()
     plane.calc_nav_pitch();
     plane.set_offset_altitude_location(plane.prev_WP_loc, plane.next_WP_loc);
     if (plane.landing.is_throttle_suppressed()) {
-       // if landing is considered complete throttle is never allowed, regardless of landing type
-       SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, 0.0);
+        // if landing is considered complete throttle is never allowed, regardless of landing type
+        SRV_Channels::set_output_scaled(SRV_Channel::k_throttle, 0.0);
     } else {
-       plane.calc_throttle();
+        plane.calc_throttle();
     }
 }
 
 void ModeAutoLand::navigate()
 {
-    if (stage == 2) {
-        // we are on final landing leg
+    switch (stage) {
+    case AutoLandStage::LOITER:
+        // check if we have arrived and completed loiter at base leg waypoint
+        if (plane.verify_loiter_to_alt(cmd[0])) {
+            stage = AutoLandStage::BASE_LEG;
+            plane.start_command(cmd[1]);
+            // setup a predictable turn angle for the base leg
+            plane.auto_state.next_turn_angle = 90;
+            plane.auto_state.crosstrack = false;
+        }
+        break;
+
+    case AutoLandStage::BASE_LEG:
+        // check if we have reached the final approach waypoint
+        if (plane.verify_nav_wp(cmd[1])) {
+            stage = AutoLandStage::LANDING;
+            plane.start_command(cmd[2]);
+        }
+        break;
+
+    case AutoLandStage::LANDING:
         plane.set_flight_stage(AP_FixedWing::FlightStage::LAND);
-        plane.verify_command(cmd[stage]);
+        plane.verify_command(cmd[2]);
+        // make sure we line up
+        plane.auto_state.crosstrack = true;
+        break;
+    }
+}
+
+/*
+  Takeoff direction is initialized after arm when sufficient altitude
+  and ground speed is obtained, then captured takeoff direction +
+  offset used as landing direction in AUTOLAND
+*/
+void ModeAutoLand::check_takeoff_direction()
+{
+#if HAL_QUADPLANE_ENABLED
+    // we don't allow fixed wing autoland for quadplanes
+    if (quadplane.available()) {
         return;
     }
+#endif
 
-    // see if we have completed the leg
-    if (plane.verify_nav_wp(cmd[stage])) {
-        stage++;
-        plane.prev_WP_loc = plane.next_WP_loc;
-        plane.auto_state.next_turn_angle = 90;
-        plane.start_command(cmd[stage]);
-        plane.auto_state.crosstrack = true;
-        plane.auto_state.next_wp_crosstrack = true;
+    if (plane.takeoff_state.initial_direction.initialized) {
+        return;
+    }
+    //set autoland direction to GPS course over ground
+    if (plane.control_mode->allows_autoland_direction_capture() &&
+        plane.is_flying() &&
+        hal.util->get_soft_armed() &&
+        plane.gps.ground_speed() > GPS_GND_CRS_MIN_SPD) {
+        set_autoland_direction();
```

Comment:
```suggestion
        set_autoland_direction(plane.gps.ground_course() + landing_dir_off);
```
If you passing in the heading the `set_autoland_direction` can be used for both this and the compass case saving some flash.

Reply:
done